### PR TITLE
Deprecate returning extras with grads

### PR DESCRIPTION
**Pull request by carmocca**
## What does this PR do?

Fixes https://github.com/PyTorchLightning/pytorch-lightning/pull/7882#discussion_r651348592
cc: @kandluis @simran2905

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [n/a] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

### CODE REVIEW

File: pytorch_lightning/trainer/connectors/logger_connector/result.py [hunk 1]
```diff
@@ -279,9 +282,16 @@ def extra(self, extra: Mapping[str, Any]) -> None:
 
         def check_fn(v):
             if v.grad_fn is not None:
-                raise MisconfigurationException(f'You returned a tensor with `grad_fn`. The extra values are {extra}')
+                # raise MisconfigurationException(f'You returned a tensor with `grad_fn`. The extra values are {extra}')
```
Comment (kandluis) [inline pytorch_lightning/trainer/connectors/logger_connector/result.py:285]: Do we need to update the `metrics(...)` function to also detach when returning the collection?

```
            # populate logging metrics
            if result_metric.meta.logger:
                metrics[MetricSource.LOG][forked_name] = value.detach()

            # populate callback metrics. callback metrics don't take `_step` forked metrics
            if self.training or result_metric.meta.on_epoch and not on_step:
                metrics[MetricSource.CALLBACK][name] = value.detach()
                metrics[MetricSource.CALLBACK][forked_name] = value.detach()
```

I'm not super familiar with the old code, so I'm not sure if automatically detaching was something we did only for `extra` or for everything else as well? I suspect it must have been for everything since we have some tests internally that `deepcopy` the LMs and this does not work when not detaching the tensors.

Comment (simran2905) [reply pytorch_lightning/trainer/connectors/logger_connector/result.py:285]: I think we were removing the _results parameter from the pl_module before deepcopying which doesn't exist anymore. Might just need to see what its saved now as

Comment (kandluis) [reply pytorch_lightning/trainer/connectors/logger_connector/result.py:285]: Yep, updated the corresponding code internally. That addresses our quantization piece, but I think this change is still needed for other libraries. There's an internal test that fails because we do a deepcopy in model_checkpoint.py:

```
# from model_checkpoint.py

def _monitor_candidates(self, trainer: 'pl.Trainer', epoch: int, step: int) -> Dict[str, _METRIC]:
        monitor_candidates = deepcopy(trainer.logger_connector.callback_metrics)
        monitor_candidates.update(epoch=epoch, step=step)
        return monitor_candidates
```

The callback metrics dictionary there includes `train_loss` and `train_loss_step` which both have `require_grad=True` at that point (so deepcopy is not possible).

The LM seems to do the right thing though, since all it does is return a single loss tensor in `training_step`.

Comment (carmocca) [reply pytorch_lightning/trainer/connectors/logger_connector/result.py:285]: Are you using `self.log(..., enable_graph=True)`?

**Review by awaelchli — APPROVED** [review]
LGTM
you have some commented lines which would need removal :)

**Review by justusschock — APPROVED** [review]

File: pytorch_lightning/trainer/connectors/logger_connector/result.py (hunk 1 above)
Comment (kandluis) [reply pytorch_lightning/trainer/connectors/logger_connector/result.py]: Ah, you're right! Yes, the LM is using this option :) Will try removing it now.

Comment (carmocca) [reply pytorch_lightning/trainer/connectors/logger_connector/result.py]: Great. Merging this. If there's any further issue with `enable_graph`, we can deal with it separately

**Review by tchaton — APPROVED** [review]
LGTM !
